parser/src/parse.rs

```diff
@@ -0,0 +1,242 @@
+/// This function was adopted from [snailquote](https://docs.rs/snailquote/latest/snailquote/).
+///
+/// # Details
+///
+/// Parses a single or double quoted string and interprets escape sequences such as
+/// '\n', '\r', '\'', etc.
+///
+/// Supports raw strings prefixed with `r` or `R` in which case all escape sequences are ignored.///
+///
+/// The full set of supported escapes between quotes may be found below:
+///
+/// | Escape     | Code       | Description                              |
+/// |------------|------------|------------------------------------------|
+/// | \a         | 0x07       | Bell                                     |
+/// | \b         | 0x08       | Backspace                                |
+/// | \v         | 0x0B       | Vertical tab                             |
+/// | \f         | 0x0C       | Form feed                                |
+/// | \n         | 0x0A       | Newline                                  |
+/// | \r         | 0x0D       | Carriage return                          |
+/// | \t         | 0x09       | Tab                                      |
+/// | \\         | 0x5C       | Backslash                                |
+/// | \?         | 0x??       | Question mark                            |
+/// | \"         | 0x22       | Double quote                             |
+/// | \'         | 0x27       | Single quote                             |
+/// | \`         | 0x??       | Backtick                                 |
+/// | \xDD       | 0xDD       | Unicode character with hex code DD       |
+/// | \uDDDD     | 0xDDDD     | Unicode character with hex code DDDD     |
+/// | \UDDDDDDDD | 0xDDDDDDDD | Unicode character with hex code DDDDDDDD |
+/// | \DDD       | 0DDD       | Unicode character with octal code DDD    |
+///
+/// # Errors
+///
+/// The returned result can display a human readable error if the string cannot be parsed as a
+/// valid quoted string.
+///
+
+pub fn parse_string(s: &str) -> Result<String, ParseError> {
+    let mut in_single_quotes = false;
+    let mut in_double_quotes = false;
+    let mut in_raw_string = false;
+
+    let mut chars = s.chars().enumerate();
+    let mut res = String::with_capacity(s.len());
+
+    while let Some((idx, c)) = chars.next() {
+        let in_quotes = in_single_quotes || in_double_quotes;
+
+        if !in_quotes && (c == 'r' || c == 'R') {
+            in_raw_string = true;
+            continue;
+        } else if c == '\\' && !in_raw_string {
+            if in_quotes {
+                match chars.next() {
+                    None => {
+                        return Err(ParseError::InvalidEscape {
+                            escape: format!("{}", c),
+                            index: idx,
+                            string: String::from(s),
+                        });
+                    }
+                    Some((idx, c2)) => {
+                        let s = match c2 {
+                            'a' => String::from("\u{07}"),
```

Comment:
Not very happy about these `String::from` - seems a bit unnecessary, but haven't figured out a better way. On the other hand I don't expect parsing to be executed too frequently, so it might be fine for now.

The original code was returning a character here, but now there are cases when we need to return multiple characters, so I changed it to a string, because we can use `push_str` to append pieces to result.